src/peft/tuners/road/layer.py

```diff
@@ -0,0 +1,398 @@
+    yₙ = x₀ * α * sinθ + xₙ * α * cosθ
+
+    The scales α and angles θ are learned for each pair of elements and, moreover, each of the 4 instances in the
+    rotation matrix may actually be different (when using variant 2 or 4).
+
+    Note that instead of using two consecutive elements x₀ x₁ we first split the whole vector into groups and pair
+    elements from the first with the second half of the same group, which allows for more efficient inference implementation.
+
+    The adapter needs to only store the angles θ and scales α, rather than the full matrix R and the inference
+    implementation only needs to do elementwise vector multiplications.
+
+    For merging the weights, we make use of the following formula: R @ (W @ x + b) = (R @ W) @ x + R @ b. The lhs part
+    is how it is used in unmerged state (using efficient elementwise implementation instead of matrix multiplication)
+    and the rhs part is how it is used in merged state where (R @ W) becomes the new weight matrix and R @ b becomes
+    the new bias.
+
+    """
+
+    adapter_layer_names: tuple[str, ...] = ("road_theta", "road_alpha")
+
+    def __init__(self, base_layer: nn.Module, ephemeral_gpu_offload: bool = False, **kwargs) -> None:
+        self.base_layer = base_layer
+        self.variant = {}
+        self.group_size = {}
+        self.road_theta = nn.ParameterDict({})
+        self.road_alpha = nn.ParameterDict({})
+
+        self._disable_adapters = False
+        self.merged_adapters = []
+
+        base_layer = self.get_base_layer()
+        if isinstance(base_layer, nn.Linear):
+            in_features, out_features = base_layer.in_features, base_layer.out_features
+        else:
+            in_features, out_features = None, None
+            warnings.warn(
+                f"Unsupported layer type '{type(base_layer)}' encountered, proceed at your own risk.", UserWarning
+            )
+        self.in_features = in_features
+        self.out_features = out_features
+
+    @property
+    def _available_adapters(self) -> set[str]:
+        return {*self.road_theta}
+
+    def update_layer(
+        self,
+        adapter_name,
+        variant,
+        group_size,
+        init_weights,
+    ):
+        # collect the kwargs
+        kwargs = locals().copy()
+        del kwargs["self"]
+
+        self.variant[adapter_name] = variant
+        self.group_size[adapter_name] = group_size
+
+        # Actual trainable parameters
+        if variant == RoadVariant.ROAD_1:
+            size = self.out_features // 2
+        elif variant == RoadVariant.ROAD_2:
+            size = self.out_features
+        elif variant == RoadVariant.ROAD_4:
+            size = self.out_features * 2
+        else:
+            raise ValueError(f"Unsupported variant {variant} for RoadLayer. Supported variants are 1, 2, and 4.")
+        self.road_theta[adapter_name] = nn.Parameter(torch.rand(size))
+        self.road_alpha[adapter_name] = nn.Parameter(torch.rand(size))
```

Comment:
WDYT about initializing this with `torch.empty` and then in `reset_road_parameters`, in case that `init_weights=False`, fill them with `rand`? That way, if `init_weights=True` (which is most common), we don't double-initialize the values, making the initialization more efficient.

Reply:
When passing `init_weights=False`, the values should be randomly initialized in `reset_road_parameters` (which can be renamed to `reset_parameters`) to avoid this.